proxyfinder.go

```diff
@@ -85,34 +89,36 @@ func (pf *ProxyFinder) findProxyForRequest(req *http.Request) (*url.URL, error)
 		log.Printf(`[%d] %s %s via "DIRECT" (not connected)`, id, req.Method, req.URL)
 		return nil, nil
 	}
-	s, err := pf.runner.FindProxyForURL(req.URL)
+	str, err := pf.runner.FindProxyForURL(req.URL)
 	if err != nil {
 		return nil, err
 	}
-	log.Printf("[%d] %s %s via %q", id, req.Method, req.URL, s)
-	ss := strings.Split(s, ";")
-	if len(ss) > 1 {
-		log.Printf("[%d] Warning: ignoring all but first proxy in %q", id, s)
-	}
-	trimmed := strings.TrimSpace(ss[0])
-	if trimmed == "DIRECT" {
-		return nil, nil
-	}
-	var host string
-	n, err := fmt.Sscanf(trimmed, "PROXY %s", &host)
-	if err == nil && n == 1 {
-		// The specified proxy should contain both a host and a port, but if for some reason
-		// it doesn't, assume port 80. This needs to be made explicit, as it eventually gets
-		// passed to net.Dial, which also requires a port.
-		proxy := &url.URL{Host: host}
-		if proxy.Port() == "" {
-			proxy.Host = net.JoinHostPort(host, "80")
+	arr := strings.Split(str, ";")
+	for _, elem := range arr {
+		fields := strings.Fields(strings.TrimSpace(elem))
+		if len(fields) == 1 && fields[0] == "DIRECT" {
+			log.Printf("[%d] %s %s via %q", id, req.Method, req.URL, elem)
+			return nil, nil
+		} else if len(fields) == 2 && fields[0] == "PROXY" {
+			// The specified proxy should contain both a host and a port, but if for
+			// some reason it doesn't, assume port 80. This needs to be made explicit,
+			// as it eventually gets passed to net.Dial, which also requires a port.
+			proxy := &url.URL{Host: fields[1]}
+			if proxy.Port() == "" {
+				proxy.Host = net.JoinHostPort(proxy.Host, "80")
+			}
+			if pf.blocked.contains(proxy.Host) {
+				log.Printf("[%d] Skipping bad proxy: %q", id, proxy.Host)
+				continue
+			}
+			log.Printf("[%d] %s %s via %q", id, req.Method, req.URL, elem)
+			return proxy, nil
 		}
-		return proxy, nil
-	}
-	if strings.HasPrefix(trimmed, "SOCKS ") {
-		return nil, errors.New("Alpaca does not yet support SOCKS proxies")
-	} else {
-		return nil, errors.New("Couldn't parse PAC response")
+		log.Printf("[%d] Coudln't parse proxy: %q", id, elem)
```

Comment:
typo: Couldn't

Also, this will log for every request if the proxy string is malformed. it probably only needs to be logged once.

Reply:
Fixed the typo. I think you're right that this is a bit spammy, but I might leave this here because it's a bit too much work to go and track which malformed proxy strings we've encountered.

Also, I think the most likely case that this would happen is an unsupported proxy scheme (e.g. SOCKS), in which case Alpaca is going to be unsuitable for the user. Alpaca should probably complain loudly in this case...

Reply:
fair enough. perhaps you should log it in ALL CAPS as it is louder.